Add support for Dual{Complex}

MikaelSlevinsky opened the pull request:
Dear dualists (duellers?)

We would like to use DualNumbers in ApproxFun and dependent packages such as SingularIntegralEquations as a requirement. In ApproxFun, we create expansions of functions in numerous spectral bases (Chebyshev, Fourier, Legendre, Jacobi, etc...) and solve ordinary and partial differential equations in these bases as well.

One particular use of dual numbers will be in automatic differentiation to compute normal derivatives of incident waves to solve elliptic PDEs like the Helmholtz equation with Neumann boundary conditions in SingularIntegralEquations.

To make the transition possible, DualNumbers must be extended to work on Complex data types as well. This required a minor rewrite, and for the most part an improvement.

I would be happy to see this merged, though I can be patient given dependencies on DualNumbers that would break.

If merged, this would close #13, #23 (as a patch), and #24.

mlubin commented on test/automatic_differentiation_test.jl:
```diff
@@ -1,52 +1,52 @@
     it = x
     while abs(it*it - x) > 1e-13
-        it = it - (it*it-x)/(2it)
+        it = (it+x/it)/2
```
Any reason for this change?

MikaelSlevinsky replied:
It's just two fewer operations https://en.wikipedia.org/wiki/Square_root#Computation

mlubin commented:
This looks potentially very useful, thanks! We've been waiting for an application where this was needed, do you have an example or small write up that explains why dual numbers make sense with complex element types? What about ``Complex{Dual}``?

Please deprecate the methods which are being removed, also use ``@Base.deprecate_binding`` for the renaming of ``du``.

mlubin commented:
CC @goretkin

MikaelSlevinsky commented:
I guess this breaks with 0.3 as well.

Some toy examples are [here](https://github.com/ApproxFun/ApproxFun.jl/blob/feat-dual-compat/examples/Dual%20Numbers.jl), and with this PR, they carry over naturally to automatic differentiation of functions of a complex variable.

I haven't thought about `Complex{Dual}`, but since Julia's complex is actually `Complex{T<:Real}`, this would only be possible with `Dual <: Real`, which would be odd given `Dual{Complex{Float64}}`.

mlubin commented:
I'm okay with breaking 0.3, just be sure to update REQUIRE and the travis settings.

goretkin commented:
I haven't followed along too carefully, but I imagine that ApproxFun needs to take some derivatives through, for example, a DFT.

You could define a DFT that doesn't use any complex arithmetic, but the way the DFT is usually expressed relies on complex arithmetic.

This is from a while back, but here's a sort-of clear example https://github.com/goretkin/ComplexDualNumbers.jl/blob/master/examples/dft_autodiff.jl

I defined two functions `spec` and `spec_nocomplex` which assigns some scalar to the spectrum of its inputs. They're the same, but internally one uses complex arithmetic. The point is that it's a R^n->R function, and you should be able to take its derivative.

goretkin commented:
For me, the biggest question is still what to do about Complex{Dual} and Dual{Complex}. Current definitions aside, it _does_ make sense to interpret a "Complex Dual" number both as a Complex number with Dual parts, and as a Dual number with Complex parts. The same goes for Dual Quaternions.

jrevels commented on src/dual.jl:
```diff
@@ -1,68 +1,67 @@
+
+immutable Dual{T<:ReComp} <: Number
+    value::T
+    dual::T
```
First of all, +1 to this, thanks for writing it up. I'm really interested in seeing this in action.

Can we use `epsilon` instead of `dual` for the `ɛ` component? I feel like it would be clearer, and also less confusing for users transitioning to the new version of DualNumbers.jl (since before, `dual` was the name of the vectorized constructor).

jrevels commented on src/dual.jl:
```diff
@@ -1,68 +1,67 @@
 
 typealias Dual128 Dual{Float64}
 typealias Dual64 Dual{Float32}
 typealias DualPair Dual
```
I know this isn't exactly in the scope of this PR, but is there any reason to keep around `DualPair` (I'm not sure why it exists in the first place)?

jrevels commented on src/dual.jl:
```diff
@@ -1,68 +1,67 @@
+Dual64(x::Float32, y::Float32) = Dual{Float32}(x, y)
+Dual64(x::Real, y::Real) = Dual64(float32(x), float32(y))
+Dual64(z) = Dual64(value(z), dual(z))
+=#
```
I feel like we just toss these constructors, or at the very least make them typealiases instead of functions. @mlubin thoughts?

MikaelSlevinsky commented:
Thanks for taking a look!

@mlubin, I deprecated the binding for `du` and modified the require and travis build file.

@jrevels, I tossed the typealias constructors, reverted `dual` to `epsilon` and re-added `dual` as a vectorized constructor.

Now, the type `Dual` contains a `value` and an `epsilon`, the display is an `ɛ` and the method `epsilon()` gets the `epsilon` component of a Dual. Very consistent. (The main issue was the use of `real(::Dual)` really.)

@goretkin, @dlfivefifty showed me that the DFT/DCT of `Vector{Dual}` can just be the `dual(DCT(value),DCT(epsilon))`, so `Dual{Complex}` satisfies our needs.

mlubin commented on .travis.yml:
```diff
@@ -1,10 +1,10 @@
 julia:
-  - 0.3
-  - 0.4
+  - release
```
This should be ``0.4``

MikaelSlevinsky commented:
@mlubin, fixed!

mlubin commented on src/dual.jl:
```diff
@@ -1,68 +1,57 @@
-
+typealias Dual64  Dual{Float32}
+typealias Dual32  Dual{Float16}
+typealias DualComplex256 Dual{Complex128}
```
Do you find these useful or is this just for consistency?

MikaelSlevinsky replied:
I don't know if I'll use them more often than the generic constructor, but I guess their purpose is to mirror `Float64` and `Complex128` etc...

mlubin commented:
The name ``value`` seems too generic to export, it will lead to conflicts with other packages. Is there some other existing method in Base which we can extend?

MikaelSlevinsky commented:
The names I can think of are:
```julia
num
values # assumes there are many
var
base
read # this is a stretch
```

dlfivefifty commented:
How about “diag”, since dual(a,b) is equivalent to

[a b; 0 a]


> On 25 Nov 2015, at 10:41 PM, Richard Mikael Slevinsky <notifications@github.com> wrote:
> 
> The names I can think of are:
> 
> num
> values # assumes there are many
> var
> base
> read # this is a stretch
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-159580870>.
>

mlubin commented:
What about defining:
```
real{T<:Real}(x::Dual{T}) = x.value
```
In addition to avoiding needless deprecations in existing code, ``real`` just seems like the mathematically reasonable description of the first component of a dual number when the components themselves are real, by analogy to the imaginary numbers. We don't need to generalize this away... most users will be interacting with ``Dual{Float64}``, so I think it's a useful concept to keep in place. For the special case of complex-valued components, maybe an unexported ``DualNumbers.value`` or ``diag`` or one of the other suggestions?

mlubin commented:
Another option is to define
```julia
getindex(x::Dual, idx) = (idx == 1) ? x.value : x.epsilon # plus some error checking
```
and use ``x[1]`` for the first component and ``x[2]`` for the second. There's a runtime penalty though, so this would be the cute way but not necessarily the fastest way to get the components.

dlfivefifty commented:
How about just using “x.value” and “x.epsilon”?


> On 26 Nov 2015, at 12:15 PM, Miles Lubin <notifications@github.com> wrote:
> 
> Another option is to define
> 
> getindex(x::Dual, idx) = (idx == 1) ? x.value : x.epsilon # plus some error checking
> and use x[1] for the first component and x[2] for the second. There's a runtime penalty though, so this would be the cute way but not necessarily the fastest way to get the components.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-159772606>.
>

mlubin commented:
I don't see a reason to remove the current accessors for the real case, but if the dot syntax helps for the complex case then I'm not opposed.

jrevels commented:
+1 for unexported `DualNumbers.value`, and just letting people use `x.value`/`x.epsilon` otherwise.

I'm also in favor of the `real` definition @mlubin proposed, with the caveat that there should probably be an explicit error message for `real(::Dual{T<:Complex})`.

dlfivefifty commented:
real(x::Dual{T<:Complex}) should return dual(real(x.value),real(x.epsilon))



> On 26 Nov 2015, at 4:58 PM, Jarrett Revels <notifications@github.com> wrote:
> 
> +1 for unexported DualNumbers.value, and just letting people use x.value/x.epsilon otherwise.
> 
> I'm also in favor of the real definition @mlubin <https://github.com/mlubin> proposed, with the caveat that there should probably be an explicit error message for real(::Dual{T<:Complex}).
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-159813237>.
>

jrevels commented:
> real(x::Dual{T<:Complex}) should return dual(real(x.value),real(x.epsilon))

Isn't there an expectation that `real(::Number)` to return a `T<:Real`? This expectation would be violated if `real(::Dual)` were to ever return a `Dual`.

dlfivefifty commented:
I think “real” here is short for “realpart”, like “imag” is short for “imagpart”.  So has nothing to do with the type Real.    

I’d expect Real(::Number) to return a Real.  


> On 26 Nov 2015, at 5:07 PM, Jarrett Revels <notifications@github.com> wrote:
> 
> real(x::Dual{T<:Complex}) should return dual(real(x.value),real(x.epsilon))
> 
> Isn't there an expectation that real(::Number) to return a T<:Real? This expectation would be violated if real(::Dual) were to ever return a Dual.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-159815048>.
>

MikaelSlevinsky commented:
:+1: to unexported `value`. (For frequent use, it's a one-liner to `import DualNumbers: value` anyway.)

If we choose to be mathematically rigorous, shouldn't `real{T<:Real}(z::Dual{T}) = z` and `real{T<:Complex}(z::Dual{T})` be undefined (or throw an error) since `Re(z)` is not complex differentiable? This makes deprecation complicated, however.

The same holds for `conj` and `imag`, but `imag{T<:Real}(z::Dual{T})` should return `zero(z)`.

One option is to add `realdual` and `imagdual` like `conjdual` for the other notions that are being discussed.

dlfivefifty commented:
Doesn't dual tell the angle of differentiation?  So it’s not about complex differentiability, but differentiability at an angle?



> On 26 Nov 2015, at 9:34 PM, Richard Mikael Slevinsky <notifications@github.com> wrote:
> 
>  to unexported value. (For frequent use, it's a one-liner to import DualNumbers: value anyway.)
> 
> If we choose to be mathematically rigorous, shouldn't real{T<:Real}(z::Dual{T}) = z and real{T<:Complex}(z::Dual{T}) be undefined (or throw an error) since Re(z) is not complex differentiable? This makes deprecation complicated, however.
> 
> The same holds for conj and imag, but imag{T<:Real}(z::Dual{T}) should return zero(z).
> 
> One option is to add realdual and imagdual like conjdual for the other notions that are being discussed.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-159872145>.
>

MikaelSlevinsky commented:
The dual part returns the epsilon component of z multiplied by f'(z.value), so:
```julia
julia> z = Dual(1.0,1.0+im)
1.0 + 0.0im  +  1.0 + 1.0imɛ

julia> exp(z)
2.718281828459045 + 0.0im  +  2.718281828459045 + 2.718281828459045imɛ
```
Doesn't this mean that the function must be complex differentiable?

dlfivefifty commented:
No I think it's just

lim (f(z.value + h*z.epsilon) - f(z.value))/h

When f is complex differentiable, this reduces to f'(z.value)z.epsilon, but that's just a special case.

Sent from my iPhone

> On 26 Nov 2015, at 9:55 PM, Richard Mikael Slevinsky <notifications@github.com> wrote:
> 
> The dual part returns the epsilon component of z multiplied by f'(z.value), so:
> 
> julia> z = Dual(1.0,1.0+im)
> 1.0 + 0.0im  +  1.0 + 1.0imɛ
> 
> julia> exp(z)
> 2.718281828459045 + 0.0im  +  2.718281828459045 + 2.718281828459045imɛ
> Doesn't this mean that the function must be complex differentiable?
> 
> —
> Reply to this email directly or view it on GitHub.
>

MikaelSlevinsky commented:
That's a really good point, and we'll need `abs` for normal derivatives in SingularIntegralEquations. This allows dual definitions of `real`, `imag`, `conj`, `abs`, and `abs2` and potentially more later. Now the dual part of `abs` changes as you vary the direction (epsilon component). Bonus, since abs is real, the epsilon is also real:
```julia
julia>  using DualNumbers

julia> z = Dual(1.0+1.0im,1.0)
1.0 + 1.0im  +  1.0 + 0.0imɛ

julia> abs(z), sqrt(2) + 1/sqrt(2)*ɛ
(1.4142135623730951 + 0.7071067811865475ɛ,1.4142135623730951 + 0.7071067811865475ɛ)

julia> z = Dual(1.0+1.0im,cis(π/4))
1.0 + 1.0im  +  0.7071067811865476 + 0.7071067811865475imɛ

julia> abs(z), sqrt(2) + 2/sqrt(2)^2*ɛ
(1.4142135623730951 + 0.9999999999999998ɛ,1.4142135623730951 + 0.9999999999999998ɛ)

julia> z = Dual(1.0+1.0im,1.0im)
1.0 + 1.0im  +  0.0 + 1.0imɛ

julia> abs(z), sqrt(2) + 1/sqrt(2)*ɛ
(1.4142135623730951 + 0.7071067811865475ɛ,1.4142135623730951 + 0.7071067811865475ɛ)

```

goretkin commented:
It would be nice if 

```
julia> z = Dual(1.0+1.0im,1.0im)
1.0 + 1.0im  +  0.0 + 1.0imɛ
```

instead printed as 
```
julia> z = Dual(1.0+1.0im,1.0im)
1.0 + 1.0im  +  0.0ɛ + 1.0imɛ
```

or

```
julia> z = Dual(1.0+1.0im,1.0im)
1.0 + 1.0im  +  (0.0 + 1.0im)ɛ
```

MikaelSlevinsky commented:
Yes I agree, it's fixed now and gives:
```julia
julia> using DualNumbers

julia> one(DualComplex256)
1.0 + 0.0im + (0.0 + 0.0im)ɛ

julia> zeros(DualComplex256,2)
2-element Array{DualNumbers.Dual{Complex{Float64}},1}:
 0.0+0.0im + (0.0+0.0im)ɛ
 0.0+0.0im + (0.0+0.0im)ɛ
```
This way, the numbers can be copied and pasted and create new dual numbers.

mlubin commented:
This basically looks good to merge if you could add back the definition ``real{T<:Real}(x::Dual{T}) = x.value``

MikaelSlevinsky commented:
The issue is that it breaks with the limit definition which implies `real(x::Dual) = Dual(real(x.value),real(x.epsilon))`. I don't know what to say! I added a comment on the readme stating the more general limit definition and how the derivative is a consequence for complex differentiable functions.

mlubin commented:
I'm personally okay with a small (documented) inconsistency here for the complex case for the benefit of the ``Dual{Float64}`` case. Will this cause any problems when differentiating complex code?

goretkin commented:
In case it is helpful, the way I see it is

in complex.jl, `real` means the non-imaginary part of the number (first definition)
in DualNumbers.jl, `real` has meant the non-epsilon part of the number (second definition).

When you have `Complex{Real}` or `Dual{Real}`, these two meanings are non-contradictory, and both return a `Real`.  But if you have Dual{Complex}, the meanings are in conflict. The definition that @MikaelSlevinsky says must follow is the first definition. @mlubin wants the second definition back, but only for Dual{Real}.

dlfivefifty commented:
And I think Mikaels argument is that if real is treated as a regular function (like exp, etc.), then it should have the behaviour specified by taking the limit.

So even without complex numbers, the behaviour of real was inconsistent.

Sent from my iPhone

> On 1 Dec 2015, at 7:43 AM, Gustavo Goretkin <notifications@github.com> wrote:
> 
> In case it is helpful, the way I see it is
> 
> in complex.jl, real means the non-imaginary part of the number (first definition)
> in DualNumbers.jl, real has meant the non-epsilon part of the number (second definition).
> 
> When you have Complex{Real} or Dual{Real}, these two meanings are non-contradictory, and both return a Real. But if you have Dual{Complex}, the meanings are in conflict. The definition that @MikaelSlevinsky says must follow is the first definition. @mlubin wants the second definition back, but only for Dual{Real}.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mlubin commented:
Yes, it's inconsistent in the real case with the definition of ``real(x::Float64) = x``, but that doesn't seem to have caused any trouble and is reasonably intuitive to explain. Silently changing the current behavior without any deprecation warnings is a nonstarter; there's a lot of published and unpublished packages, examples, notebooks, etc, that use this package (CC @rgiordan @lindahua). Even going through deprecations doesn't seem worth it to me, though I'm open to the argument.

dlfivefifty commented:
I would argue for adding Number and Real constructors and deprecating real(::Dual{T<:Real}):

	Number(x::Dual)=x.value
	Real{T<:Real}(x::Dual{T})=x.value
	@deprate real{T<:Real}(x::Dual{T}) Real(x)
	real{T<:Complex}(x::Dual{T})=dual(real(x.value),real(x.epsilon))

This makes updating existing code a one-character change.  

It wouldn’t be Julia without having to go through deprecations!




> On 1 Dec 2015, at 9:47 AM, Miles Lubin <notifications@github.com> wrote:
> 
> Yes, it's inconsistent in the real case with the definition of real(x::Float64) = x, but that doesn't seem to have caused any trouble and is reasonably intuitive to explain. Silently changing the current behavior without any deprecation warnings is a nonstarter; there's a lot of published and unpublished packages, examples, notebooks, etc, that use this package (CC @rgiordan <https://github.com/rgiordan> @lindahua <https://github.com/lindahua>). Even going through deprecations doesn't seem worth it to me, though I'm open to the argument.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-160786131>.
>

jrevels commented:
> I would argue for adding Number and Real constructors and deprecating real(::Dual{T<:Real}):

-1 to this. I can see what you're going for, but defining `real{T<:Real}(::Dual{T})` would automatically grant us vectorization as expected behavior for the method (which I think is nice). The `Real` conversion you defined also doesn't make as much sense in cases where it would be lossy - the existing behavior (correctly, IMO) throws an `InexactError` instead of silently throwing away numeric information:

```julia
julia> Real(1 + 0im)
1

julia> Real(1+im)
ERROR: InexactError()
 in call at essentials.jl:56

julia> Real(Dual(1, 0))
1

julia> Real(Dual(1, 1))
ERROR: InexactError()
 in call at essentials.jl:56
```

I'm also still in favor of `real{T<:Complex}(::Dual{T})` throwing an error, as returning a `Dual` from that method seems semantically inconsistent to me...`real` returning a type that's *not* `T<:Real` feels like a jarring deviation from the expectations of the existing `real` definitions in Base.

dlfivefifty commented:
The problem is the real may be used in a function that you want to apply automatic differentiation to.  For example,

f(x)=real(exp(im*x))

it seems unfortunate to forbid automatic differentiation on such functions.

Sent from my iPhone

> On 1 Dec 2015, at 10:46 AM, Jarrett Revels <notifications@github.com> wrote:
> 
> I would argue for adding Number and Real constructors and deprecating real(::Dual{T<:Real}):
> 
> -1 to this. I can see what you're going for, but defining real{T<:Real}(::Dual{T}) would automatically grant us vectorization as expected behavior for the method (which I think is nice). The Real conversion you defined also doesn't make as much sense in cases where it would be lossy - the existing behavior (correctly, IMO) throws an InexactError instead of silently throwing away numeric information:
> 
> julia> Real(1 + 0im)
> 1
> 
> julia> Real(1+im)
> ERROR: InexactError()
>  in call at essentials.jl:56
> 
> julia> Real(Dual(1, 0))
> 1
> 
> julia> Real(Dual(1, 1))
> ERROR: InexactError()
>  in call at essentials.jl:56
> I'm also still in favor of real{T<:Complex}(::Dual{T}) throwing an error, as returning a Dual from that method seems semantically inconsistent to me...real returning a type that's not T<:Real feels like a jarring deviation from the expectations of the existing real definitions in Base.
> 
> —
> Reply to this email directly or view it on GitHub.
>

jrevels commented:
> The issue is that it breaks with the limit definition which implies real(x::Dual) = Dual(real(x.value),real(x.epsilon))

> The problem is the real may be used in a function that you want to apply automatic differentiation to.  

Oh, sorry, I had forgotten about this (and the discussion surrounding it). This is actually a pretty good argument.

> Yes, it's inconsistent in the real case with the definition of real(x::Float64) = x, but that doesn't seem to have caused any trouble and is reasonably intuitive to explain

I'd argue that the reason it hasn't caused trouble in the past because nobody could use `Dual` to do autodiff on complex functions, which are much more likely to have `real` in them. This is a tougher decision than I had originally thought...

mlubin commented:
A bit of abstraction could help. We could define most of the common infrastructure using ``AbstractDual``, and then have ``immutable Dual{T<:Real} ... end`` with the current semantics and ``immutable DualComplex{T<:ReComp} ... end`` with the semantics that makes sense for differentiating complex-valued functions.

dlfivefifty commented:
Yes that's definitely a solution, though is it worth the extra complication?

The arguments for not deprecating real seem to be:

1) it requires changing existing code
2) real usually returns a real


For 2: I'd argue that real is in the same category as abs, angle, and imag which also usually return a real, but I don't think it's confusing that abs(d::Dual) returns a Dual.

For 1: I'd argue that Julia in general is constantly going through breaking changes while the ecosystem is still small, and it's better to make changes now than later.





Sent from my iPhone

> On 1 Dec 2015, at 11:16 AM, Miles Lubin <notifications@github.com> wrote:
> 
> A bit of abstraction could help. We could define most of the common infrastructure using AbstractDual, and then have immutable Dual{T<:Real} ... end with the current semantics and immutable DualComplex{T<:ReComp} ... end with the semantics that makes sense for differentiating complex-valued functions.
> 
> —
> Reply to this email directly or view it on GitHub.
>

goretkin commented:
> And I think Mikaels argument is that if real is treated as a regular function (like exp, etc.), then it should > have the behaviour specified by taking the limit.

> So even without complex numbers, the behaviour of real was inconsistent.

I don't understand this. The only (original) reason the function `real` exists is because complex numbers exist, so how was it inconsistent? The only reason someone might have had a call to `real` in some code that they want to auto-differentiate is because they had complex numbers in mind, right? You wouldn't be making a call to `DualNumbers.real` in any code that you hope to auto-diff.

dlfivefifty commented:
But this function might not be written with dual numbers in mind, and use real on a real number.  For example, a user might have

	f= z->real(exp(z))

and expect f(z::Dual)  to return a Dual whether or not z is Dual{T<:Real} or Dual{T<:Complex}.








> On 1 Dec 2015, at 11:44 AM, Gustavo Goretkin <notifications@github.com> wrote:
> 
> And I think Mikaels argument is that if real is treated as a regular function (like exp, etc.), then it should > have the behaviour specified by taking the limit.
> 
> So even without complex numbers, the behaviour of real was inconsistent.
> 
> I don't understand this. The only (original) reason the function real exists is because complex numbers exist, so how was it inconsistent? The only reason someone might have had a call to real in some code that they want to auto-differentiate is because they had complex numbers in mind, right? You wouldn't be making a call to DualNumbers.real in any code that you hope to auto-diff.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-160810417>.
>

goretkin commented:
I'm with you 100%. I just don't think it's accurate to say that the definition of `real` is inconsistent even if there were never any complex numbers. In that world, it would be okay to have `real` be an accessor method for the non-epsilon part of a dual number, because no one would be using `real` already as an accessor for the non-imaginary part of a complex number.

mlubin commented:
@dlfivefifty, another argument I'd add to that is that ``real``, as it is now, is intuitive and easy to explain. I've given a number of demos using DualNumbers and the current syntax is very clean in my opinion. Using ``Real`` to access the real component and ``epsilon`` for the epsilon component is a compromise that's not too aesthetically pleasing.

dlfivefifty commented:
It is inconsistent if you think of real as a normal function.  If you think of real as a special accessor, than it is not inconsistent.  

But I’d argue that real is more akin to abs and angle, and that the storage details of Complex should not be used.  For example, it’s possible that Complex numbers could be stored in abs/angle format:

	type Complex128
		abs::Float64
		angle::Float64
	end

If this were the case, real is not an accessor, and one would not expect it to be treated differently.


> On 1 Dec 2015, at 11:56 AM, Gustavo Goretkin <notifications@github.com> wrote:
> 
> I'm with you 100%. I just don't think it's accurate to say that the definition of real is inconsistent even if there were never any complex numbers. In that world, it would be okay to have real be an accessor method for the non-epsilon part of a dual number, because no one would be using real already as an accessor for the non-imaginary part of a complex number.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-160812382>.
>

dlfivefifty commented:
Yes that’s a good argument, but I think it would be just as intuitive as Real(x) or value(x).  Maybe instead of creating DualNumbers.value, we can create Base.value, so any other future package that exports value can do the same?


By the way, I still think defining

	Real(x)=x.value 

is OK since a dual number can be thought of as

	x.value + 0*x.epsilon

so it would be consistent with

julia> Real(1+0im)
1



> On 1 Dec 2015, at 12:00 PM, Miles Lubin <notifications@github.com> wrote:
> 
> @dlfivefifty <https://github.com/dlfivefifty>, another argument I'd add to that is that real, as it is now, is intuitive and easy to explain. I've given a number of demos using DualNumbers and the current syntax is very clean in my opinion. Using Real to access the real component and epsilon for the epsilon component is a compromise that's not too aesthetically pleasing.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/JuliaDiff/DualNumbers.jl/pull/29#issuecomment-160812852>.
>

goretkin commented:
For the record, I think the current behavior of `DualNumbers.real` is inconsistent with `Base.real`, and I don't see a way of that being reconciled. That's not what I was arguing against 
I think `value(x::Dual)` is a step in the right direction.

I think the best mathematical definition for real is real(x) = (x + conj(x))/2, which is another vote for `real(x::Dual{Real})=x`. Is there something like this for dual numbers? `real(x::Dual) = eps(epsilon * x)` is all I can think of.

dlfivefifty commented:
> 
> real returning a type that's not T<:Real feels like a jarring deviation from the expectations of the existing real definitions in Base.



Actually, it is not true that real always returns a real in Base:

```julia
julia> real([1+im 2; 3 4])
2x2 Array{Int64,2}:
 1  2
 3  4
```

jrevels commented:
> Actually, it is not true that real always returns a real in Base:

I should've been more clear, I meant that `real(::Number)` always returns a `Real` (though, as I mentioned, `real` is vectorized, so obviously whatever element-wise behavior we define will carry over to the vectorized definition). 

It seems that these definitions:

```julia
real{T<:Real}(x::Dual{T}) = x
real{T<:Complex}(x::Dual{T}) = Dual(real(value(x)), real(epsilon(x)))
```
 
 ...now have some pretty substantial arguments backing them, and I'm close to being won over. I especially feel like it would be very difficult to claim our `Dual` implementation supports complex autodiff without defining `real` properly, since `real` is sure to pop up everywhere in user code involving `Complex`. However, are we *absolutely sure* that the above definition of `real` delivers the correct results in the autodiff case? I basically believe it, but I'd need to test some edge cases to really convince myself.

I also do still share the same concerns as @mlubin - the standard nomenclature when discussing dual numbers is to refer to the non-perturbative component as the "real" component, and the function `real` as defined in Base means "give me the real component of this number."

I'm not sure we can resolve this without introducing inconsistency *somewhere*, but in the end, if inconsistency is introduced, I'd rather have that inconsistency be nominative than behavioral (i.e. I prefer distorting Base's assumption of `real` to breaking autodiff).

MikaelSlevinsky commented:
@jrevels, the dual part of the `real` function can be thought of as the dot product of the gradient of `f(x,y) = x` with `reim(epsilon(z))...`:
```julia
julia> using DualNumbers

julia> z = Dual(1.0+im,1.0)
1.0 + 1.0im + (1.0 + 0.0im)ɛ

julia> real(z)
1.0 + 1.0ɛ

julia> z = Dual(1.0+im,2.0+im)
1.0 + 1.0im + (2.0 + 1.0im)ɛ

julia> real(z)
1.0 + 2.0ɛ

julia> z = Dual(1.0+im,im)
1.0 + 1.0im + (0.0 + 1.0im)ɛ

julia> real(z)
1.0 + 0.0ɛ

```
As far as I can tell, this is correct.

Thoughts on either of these three ways forward?

1. restart with
```julia
immutable Dual{T<:ReComp}
    primal::T
    dual::T
end
```
and rename the accessors `value => primal` and `epsilon => dual`. This would break `dual`, `real` and `epsilon`.

2. Define `realdual` and `imagdual` the same sense as `absdual`, `abs2dual`, `conjdual`. As far as I can see, the original purpose for these functions is computationally equivalent to the interchange `im <=> ɛ`. I would be willing to deprecate the current DualNumbers `real` for v0.1.6 and remove the deprecation in v0.1.7

3. Keep this pull request on hold while I issue a pull request to Julia base for `value`. We have a strong case, but I fear that `Base` doesn't exist to resolve naming conflicts in external packages.

mlubin commented:
Who would've thought this would be so hard? :)

How about:
```julia
realpart{T<:Real}(x::Dual{T}) = x.value
value(x::Dual) = x.value
epsilon(x::Dual) = x.epsilon
export realpart, epsilon
```
I really don't want to give up the analogy of the first component being the real component in the case of ``Dual{Float64}``. ``realpart`` seems less confusing than ``realdual``.

MikaelSlevinsky commented:
Sounds good to me! I can make these changes if you're ok with me using your code.

mlubin commented:
Sure, go ahead

MikaelSlevinsky commented:
yet dual numbers can be generalized over any commutative ring R (real or complex numbers for instance) as the quotient ring R[X]/X^2 (so long as ɛ^2 = 0). https://en.wikipedia.org/wiki/Dual_number#Generalization

This page makes me want to have `dualpart` as well as `realpart` http://www.euclideanspace.com/maths/algebra/realNormedAlgebra/other/dualNumbers/index.htm

goretkin commented:
@MikaelSlevinsky I didn't understand whether that was in support of removing the derivative definition or not.

MikaelSlevinsky commented:
Sorry, I misread completely. That part is in the section on overriding Base functions, whereas the paragraphs above describe dual numbers more generally.

Usually, evaluating a function at a dual number returns a dual number, so what's wrong with the definition (in the context of that part of the readme)? Suggest any changes?

jrevels commented:
I don't want to hold up this PR with too much bikeshedding, but if we're leaning towards defining `realpart`/`dualpart` instead of `value`/`epsilon`, maybe it'd be better to say `valuepart`/`dualpart` (to make it less confusing when the `x.value` is `Complex`)? Otherwise, the proposed `realpart`/`dualpart`definitions seem like the right way to resolve this.

mlubin commented:
I'd much prefer ``realpart`` over ``valuepart``.

jrevels commented:
Fair enough, you have more experience demoing this stuff then I do.

Either way, +1 to both component accessors having names of the form `___part`. I like the consistency there.

mlubin commented on src/dual.jl:
```diff
@@ -1,68 +1,61 @@
+
+value(z::Dual) = z.value
+epsilon(z::Dual) = z.epsilon
+value(z::Number) = z
```
When would this definition be used?

MikaelSlevinsky replied:
I can't tell exactly but if you mean line 32: `value(z::Number) = z`: if you know `z`'s a `Number` but not necessarily a `Dual` it's good form to have a no-op.

mlubin replied:
I'd be in favor of removing the no-op until someone requests it for a good reason, but it's just a minor point.

goretkin replied:
I agree with @mlubin. I'm not aware of other definitions of `value`, but I am thinking of it as an accessor of the non-dual part of a dual number, and not as a "mathematical function" or whatever, so I don't think we need to (yet) worry about finding a meaning for other types of numbers.

MikaelSlevinsky commented:
Ok I removed `value(z::Number)` and `epsilon(z::Number)`. In the future, it may be useful to think of them as more than accessors - `Val(z)` and `Eps(z)` could be thought of as projection operators to the purely real/concrete and purely dual subspaces the same way as `Re(z)` and `Im(z)` act on complex z.

I also added the constant `imɛ` and special show methods, so now
```julia
julia> using DualNumbers

julia> z = 1.0+2.0im-3.0ɛ+4.0imɛ
1.0 + 2.0im - 3.0ɛ + 4.0imɛ

julia> ɛ
ɛ

julia> im*ɛ
imɛ

```

mlubin commented:
Looks basically ready to merge. Could you squash the commits?

MikaelSlevinsky commented:
Done! :)

mlubin commented:
Thanks!
I'll let this sit on master for a week or so before tagging. That'll give me some time to check for unintended breakage. @goretkin, we're open for documentation tweaks.

jrevels commented:
:tada:

MikaelSlevinsky commented:
Thanks again!

MikaelSlevinsky commented:
Is it time for a Pkg.tag? I'd like to push some code to SingularIntegralEquations but I want it to "require" the right version number of DualNumbers so users are unencumbered.

mlubin commented:
I'll tag tomorrow. I've been running on master for a while with no apparent issues.
